File: runtime/typescript/src/option.ts
import { DecodeCursor, Decoder, EncodeCursor, Encoder } from './index';

export type Option<T> = T | null;

export class OptionEncoder<T> implements Encoder<Option<T>>, Decoder<Option<T>> {
  private someEncoder: Encoder<T> & Decoder<T>;

  constructor(
    someEncoder: Encoder<T> & Decoder<T>,
  ) {
    this.someEncoder = someEncoder;
  }

  baseLength = () => 1 + this.someEncoder.baseLength();

  scratchLength(value: Option<T>): number {
    if (value !== null) {
      return this.someEncoder.scratchLength(value);
    } else {
      return 0;
    }
  }

  encode(cursor: EncodeCursor, value: Option<T>) {
    if (value !== null) {
      cursor.buffer.setUint8(cursor.base(1), 1);
      this.someEncoder.encode(cursor, value);
    } else {
      cursor.buffer.setUint8(cursor.base(1), 0);
    }
  }

  decode(cursor: DecodeCursor): Option<T> {
    let variant = cursor.buffer.getUint8(cursor.base(1));

    if (variant == 0) {
      return null;
    } else if (variant == 1) {
      return this.someEncoder.decode(cursor);
    } else {
      throw "Failed to decode Option - invalid variant tag";
    }
  }
}

export class OptionLazyDecoder<T> implements Decoder<Option<T>> {
  private someDecoder: Decoder<T>;

  constructor(
    someDecoder: Decoder<T>,
  ) {
    this.someDecoder = someDecoder;
  }

  baseLength = () => 1 + this.someDecoder.baseLength();

  decode(cursor: DecodeCursor): Option<T> {
    let variant = cursor.buffer.getUint8(cursor.base(1));

    if (variant == 0) {
      return null;
    } else if (variant == 1) {
      return this.someDecoder.decode(cursor);
    } else {
      throw "Failed to decode Option - invalid variant tag";
    }
  }
}

export const ProtoOption = <T>(someEncoder: Encoder<T> & Decoder<T>) => new OptionEncoder(someEncoder);
export const ProtoOptionLazy = <T>(someDecoder: Decoder<T>) => new OptionLazyDecoder(someDecoder);

File: runtime/rust/src/option.rs
use crate::{
    BaseLen, Compatible, Decode, DecodeCursor, DecodeError, DecodeResult, Encode, EncodeCursor,
    Lazy, Owned,
};

impl<T: Owned> Owned for Option<T> {
    type Lazy<'a> = Option<T::Lazy<'a>>;

    fn lazy_to_owned(lazy: Self::Lazy<'_>) -> DecodeResult<Self> {
        match lazy {
            Some(lazy) => Ok(Some(T::lazy_to_owned(lazy)?)),
            None => Ok(None),
        }
    }
}

impl<'a, T: Lazy<'a>> Lazy<'a> for Option<T> {
    type Owned = Option<T::Owned>;
}

impl<T, U: Compatible<T>> Compatible<Option<T>> for Option<U> {}

impl<T: BaseLen> BaseLen for Option<T> {
    const BASE_LEN: usize = 1 + T::BASE_LEN;
}

impl<T: Encode> Encode for Option<T> {
    fn scratch_len(&self) -> usize {
        match self {
            Some(some) => some.scratch_len(),
            None => 0,
        }
    }

    fn encode(&self, cursor: &mut EncodeCursor) {
        match self {
            Some(some) => {
                cursor.base(1)[0] = 1;
                some.encode(cursor);
            }
            None => {
                cursor.base(1)[0] = 0;
                cursor.base(T::BASE_LEN);
            }
        }
    }
}

impl<'a, T: Decode<'a>> Decode<'a> for Option<T> {
    fn decode(cursor: &DecodeCursor<'a>) -> DecodeResult<Self> {
        let variant = cursor.base(1)[0];
        if variant > 1 {
            return Err(DecodeError);
        }

        let is_some = variant == 1;
        if is_some {
            Ok(Some(T::decode(cursor)?))
        } else {
            cursor.advance(T::BASE_LEN);
            Ok(None)
        }
    }
}
